Add MetricSender wrapper class to monitoring lib

zgalor opened the pull request:
Continuing work done in https://github.com/openshift/openshift-tools/pull/1178.

- **MetricSender** class that wraps **HawkSender** and **ZaggSender**. Metrics are sent to each sender that's configured "**active**" ( by default Zagg is active. Hawk is inactive)

- config file: **metric_sender.yaml** (replacing zagg_client.yaml and adding hawk connectivity options).

- cli's for shell scripts:

1. **ops-metric-client** (replacing ops-zagg-client).
2. **ops-metric-pcp-client** (replacing ops-zagg-pcp-client)

- Following cli's were not replaced (used in ansible modules, and docker scripts):

1. ops-zagg-hearbet-processor
2. ops-zagg-hearbeter
3. ops-zagg-metric-processor

- Scripts in monitoring and cloud libraries were updated to use new MetricSender, and tested.

- ansible playbooks and docker shell scripts not updated and still use ops-zagg-client

- tags are not yet being sent to hawkular. We need to define the way we want to send them and add them as another commit or even a different PR

yaacov committed:
Add a hawkular configuration, client and sender classes to the monitoring lib

Currently numerical metrics are sent to Zabbix.
This PR addes the library functions for sending numerical data to Hawkular.

Description
Add a wrapper code around the python hawkular client to mimic the Zabbix client.

Requirements:
zbxsend                - for UniqueMetric (If we drop zabbix, we will need to re-implement UniqueMetric)
hawkular-client-python - for low level hawkular API support

New Classes:
HawkConnection - Connection information (if inactive is True, the hawkular client will do nothing silently)
HawkClient     - A client for hawkualr metrics API
HawkSender     - High level interface for sending metrics to hawkular.

twiest commented on openshift_tools/monitoring/generic_metric_sender.py:
```diff
@@ -0,0 +1,45 @@
+            self.config = yaml.load(file(self.config_file))
+
+    def add_dynamic_metric(self, discovery_key, macro_string, macro_array, host=None, synthetic=False):
+        pass
```
As an Abstract class, should this raise if called?

If so, the other functions should be changes as well (I'm not going to add a comment for each).

zgalor replied:
Actually I want this implemented as "Do-Nothing" in the parent class so in the case of add_dynamic_metric for example which is not implemented in HawkSender, function will default to parent. This will allow generic calls to MetricSender no matter what concrete sender supports.
I will however, raise an exception in the constructor, to prevent instantiation of this abstract class. Thanks.

twiest replied:
I see what you're going for, and that's fine.

However, we DO need dynamic metric supported by the hawkular calls as well. Dynamic metrics are anything where the script needs to append a run time specific value.

So, for instance, our disk space check. When we write it, we have no idea what devices the host will have. So we write it as a dynamic metric and then pass into the dynamic metric /dev/xvda or /dev/xvdb, etc to it.

In hawkular, what I envision is that we'll simply add more tags for the dynamic items.

I think this will make more sense once you try to convert a collector script that uses dynamic items over to work with hawkular as well.

zgalor replied:
Ok let's take https://github.com/openshift/openshift-tools/blob/stg/scripts/monitoring/cron-send-disk-metrics.py#L70 for example. So you want to replace this call with add_dynamic_metric. but do we want to add anther API for adding tags, and change the script to call it or do we want the hawk_sender to infer tags from the parameters sent to add_dynamic_metric?

twiest replied:
@yaacov correct me if I'm wrong on this, but what we discussed before, is to have a separate file that would contain a key/value pair of item name and all tags that should be associated with it.

So IMO, a dynamic item would be that plus extra tags for items passed into the function.

Thoughts?

yaacov replied:
a. in hawkular all metrics are dynamic, you can invent new ids/items as you go.
`send_metrics(key='some.new.key.i.just.think.about', value=42, tags={...})` is ok.
b. the new keys can have tags:
b.1. from the `send_metrics` command it'self.
b.2. or from some `rule` in the config file, for example:

```
some.new.key.*:
   tags:
      type: 'node'
```

---

I think a nice algorithm will be:

a. use [optional] tags given by user when sending metrics:
    for example: `send_metric(key = 'cpu.usage', value = 2, tags = {type:'node', hostname: ... })`
b. use separate file to fill in missing tags if not given by the `send_metric` command.
    for example, if the config file has the lines:

```
cpu.usage:
  tags:
     units: 'cpus'
```

   the `send_metric` command will _also_ update the `units` tag

twiest commented on openshift_tools/monitoring/hawk_client.py:
```diff
@@ -0,0 +1,87 @@
+"""
+#These are not installed on the buildbot, disabling this
+#pylint: disable=no-name-in-module,unused-import,import-error
+from openshift_tools.monitoring.metricmanager import UniqueMetric, MetricManager
```
`MetricManager` isn't used in this file.

zgalor replied:
Will fix

twiest commented on openshift_tools/monitoring/hawk_client.py:
```diff
@@ -0,0 +1,87 @@
+from hawkular.metrics import HawkularMetricsClient, MetricType, Availability
+
+#This class implements rest calls. We only have one rest call implemented
+# add-metric.  More could be added here
```
`add_metric`

zgalor replied:
:+1:

twiest commented on openshift_tools/monitoring/hawk_client.py:
```diff
@@ -0,0 +1,87 @@
+
+    def __init__(self, hawk_connection):
+        # pylint doesn't know where RestAPI is
+        #pylint: disable=undefined-variable
```
Do we still need this pylint disable?

twiest commented on openshift_tools/monitoring/hawk_client.py:
```diff
@@ -0,0 +1,87 @@
+                                            tenant_id=self.hawk_conn.tenant_id,
+                                           )
+
+    def add_metric(self, unique_metric_list):
```
Since the abstract class has `synthetic` as a param, I think this needs it as well, right?

Even if it's not used for hawkular (as all are semi-synthetic already).

zgalor replied:
**add_metric** in **hawk_client** (and also in zagg_client) is called by hawk_sender.send_metrics. This sends the already stored metrics which already holds the host data determined by synthetic param in **hawk_sender.add_metric**.
**HawkClient** is not in the class hierarchy of **MetricSender**, but it also has **add_metric** method which  actually sends the metrics to hawk/zabbix. That why it's confusing. I'll rename it to **push_metrics**, i think it will be much clearer.

twiest commented on openshift_tools/monitoring/hawk_common.py:
```diff
@@ -0,0 +1,38 @@
+    # pylint: disable=too-many-arguments
+    # This now supports ssl and need a couple of extra params
+    def __init__(self, url, user, password, ssl_verify=False, debug=False, active=True):
+        self.url = url if url.startswith('http') else 'http://' + url
```
Does this work with https?

Seems like it's being hard coded to http.

zgalor replied:
I will check

yaacov replied:
checking for `url.startswith('http')` is ok is ok for http and https ...

```
url                   | self.url
----------------------|------------------------
example.com           | http://example.com
http://example.com    | http://example.com
https://example.com   | https://example.com
```

twiest replied:
Oh, I see. Of course, it's only adding http:// if the url doesn't start with http. I read the `else` as a `then`.

twiest commented on openshift_tools/monitoring/hawk_common.py:
```diff
@@ -0,0 +1,38 @@
+        self.password = password
+        self.ssl_verify = ssl_verify
+        self.debug = debug
+        self.active = active # We do not want this lib to work ansless explisitly activated
```
`ansless explisitly`  should be `unless explicitly`

twiest commented on openshift_tools/monitoring/hawk_common.py:
```diff
@@ -0,0 +1,38 @@
+        self.context = None if self.ssl_verify else ssl._create_unverified_context()
+        self.tenant_id = '_ops'
+
+HawkHeartbeat = namedtuple("HawkHeartbeat", ["templates", "hostgroups"])
```
Is this needed?

`templates` and `hostgroups` are zabbix concepts that we're hoping to not have to bring to hawkular.

zgalor replied:
I will do more cleanup once I see things working

twiest commented on openshift_tools/monitoring/hawk_sender.py:
```diff
@@ -0,0 +1,150 @@
+# vim: expandtab:tabstop=4:shiftwidth=4
+"""
+Collect metrics and send metrics to Hawk.  The data
+being send to Hawk is done using REST API using the HawkClient
```
`send` should be `sent`

twiest commented on openshift_tools/monitoring/hawk_sender.py:
```diff
@@ -0,0 +1,150 @@
+
+        return self.config['host']['name']
+
+    def _get_default_hawk_connecton(self):
```
`_get_default_hawk_connecton` should be `_get_default_hawk_connection`

twiest commented on openshift_tools/monitoring/metric_sender.py:
```diff
@@ -0,0 +1,70 @@
+    zs = HawkSender(host=HOSTNAME, hawk_connection=ZAGGCONN)
+    zs.add_heartbeat(ZAGGHEARTBEAT)
+    zs.add_zabbix_keys({ 'test.key' : '1' })
+    zs.send_metrics()
```
This example is specific to hawk sender. It should probably be changed to show how to use MetricSender.

zgalor replied:
off course

twiest commented on openshift_tools/monitoring/metric_sender.py:
```diff
@@ -0,0 +1,70 @@
+        if self.is_zagg_active:
+            self.active_senders += ZaggSender(verbose, debug)
+
+        if self.is_hawk_active:
```
same

twiest commented on openshift_tools/monitoring/metric_sender.py:
```diff
@@ -0,0 +1,70 @@
+        self.active_senders = None
+
+        super.parse_config()
+        if self.is_zagg_active:
```
Is this a function call? If so, why no `()` on the end?

zgalor replied:
No need in python, if function doesn't take parameters

twiest replied:
I could be wrong about this, but what's happening here is that since python treats functions as first class citizens, you're checking if the the function exists.

Here's an example:

``` python
>>> def blah():
...    print "blah"
... 
>>> blah
<function blah at 0x7ff99e3da758>
>>> blah()
blah
>>> 
```

Note that without the `()` then we get back the function pointer.

I'm pretty sure we want `()` on these if statements.

zgalor replied:
You're right. Got confused with ruby... :-)

twiest commented on openshift_tools/monitoring/zagg_sender.py:
```diff
@@ -151,9 +147,13 @@ def add_zabbix_keys(self, zabbix_keys, host=None, synthetic=False):
     # Allow for 6 arguments (including 'self')
     # pylint: disable=too-many-arguments
-    def add_zabbix_dynamic_item(self, discovery_key, macro_string, macro_array, host=None, synthetic=False):
+    def _add_zabbix_dynamic_item(self, discovery_key, macro_string, macro_array, host=None, synthetic=False):
```
Is there a reason not to just rename this to `add_dynamic_metric` directly instead of doing the forward method?

twiest commented on openshift_tools/monitoring/zagg_sender.py:
```diff
@@ -135,7 +131,7 @@ def add_heartbeat(self, heartbeat, host=None):
         self.unique_metrics.append(hb_metric)
 
-    def add_zabbix_keys(self, zabbix_keys, host=None, synthetic=False):
+    def _add_zabbix_keys(self, zabbix_keys, host=None, synthetic=False):
```
Is there a reason not to just rename this to `add_metric` directly instead of doing the forward method?

zgalor replied:
I will. This is temporary for backwards compatibility and for change tracking

twiest replied:
To be clear, it's not backwards compatible because of the _ at the front.  Backwards compatible would be to leave the name exactly the same.

Can you explain the change tracking part in more detail?

We use git for change tracking, so I'm not sure why this is needed.

zgalor replied:
What I meant is that I kept this so I will remember the names of the functions I need to change in the calling scripts. This is still work in progress.

twiest reviewed (changes requested):
This is really looking good, I really like how this is shaping up.

This is exactly what we're looking for. I added a few comments.

yaacov commented:
:+1:

zgalor commented:
@twiest I made the changes you requested, also ran pylint and fixed some lint warnings.
I already transformed half of the scripts in scripts/monitoring/ to use metric_sender.
Still testing. Still not sending tags.

Few questions:
1. Should scripts under https://github.com/openshift/openshift-tools/tree/stg/scripts/cloud/aws/ and https://github.com/openshift/openshift-tools/tree/stg/scripts/cloud/gcp/ also send to hawkular in addition to zabbix?
2. I see some shell scripts in https://github.com/openshift/openshift-tools/tree/stg/scripts/monitoring/ that use ops-zagg-client. I also see wide usage of ops-zagg-metric-processor and other programs that use **ZaggSender** and **ZaggClient**. Should all those be expended to send to hawkular? do you want a cli that wraps **MetricSender** class? If so this might be a bit more complex to configure. It will require a duplication of the set of zabbix params to hawkular.

zgalor commented:
Updated all cron scripts to use MetricSender. Tested all active cron scripts.

built 2 cli tools for bash scripts:
ops-metric-client (to replace ops-zagg-client) - Still having issues with heartbeat sending
ops-metric-pcp-client (to replace ops-zagg-pcp-client) - Still needs fixing.

zgalor commented:
I removed the wip label and updated PR description.
@twiest could you please review. Also let me know if you want ansible scripts and docker start.sh updated with the ability to send to hawk.

@dkorn could you please review changes made in https://github.com/zgalor/openshift-tools/tree/27bd5ec1afb0adc449e8b86bc252fe2144afcaf9/openshift_tools/monitoring package:
- https://github.com/zgalor/openshift-tools/blob/27bd5ec1afb0adc449e8b86bc252fe2144afcaf9/openshift_tools/monitoring/generic_metric_sender.py
- https://github.com/zgalor/openshift-tools/blob/27bd5ec1afb0adc449e8b86bc252fe2144afcaf9/openshift_tools/monitoring/metric_sender.py
- https://github.com/zgalor/openshift-tools/blob/27bd5ec1afb0adc449e8b86bc252fe2144afcaf9/openshift_tools/monitoring/zagg_sender.py
- https://github.com/zgalor/openshift-tools/blob/27bd5ec1afb0adc449e8b86bc252fe2144afcaf9/openshift_tools/monitoring/hawk_sender.py

Also please review:
https://github.com/zgalor/openshift-tools/blob/27bd5ec1afb0adc449e8b86bc252fe2144afcaf9/scripts/monitoring/ops-metric-client.py

zgalor committed:
Add MetricSender to monitoring lib

Add MetricSender class that wrapps ZaggSender and HawkSender and sends metrics to each sender if configured active

zgalor commented:
I split this to 3 PRs to ease review. This one contains only the addition of MetricSender class hierarchy.
@twiest  Please review.

joelddiaz commented:
@zgalor the new files being added into openshift_tools/monitoring/\* need to be packaged up in openshift_tools/python-openshift-tools.spec

RPM build errors:
    Installed (but unpackaged) file(s) found:
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/generic_metric_sender.py
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/generic_metric_sender.pyc
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/generic_metric_sender.pyo
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_client.py
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_client.pyc
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_client.pyo
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_common.py
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_common.pyc
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_common.pyo
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_sender.py
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_sender.pyc
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/hawk_sender.pyo
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/metric_sender.py
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/metric_sender.pyc
   /usr/lib/python2.7/site-packages/openshift_tools/monitoring/metric_sender.pyo

zgalor commented:
@joelddiaz Done

joelddiaz commented on openshift_tools/monitoring/hawk_client.py:
```diff
@@ -0,0 +1,91 @@
+    print client.add_metric(ml)
+"""
+# The hawkular client
+from hawkular.metrics import HawkularMetricsClient, MetricType
```
@zgalor this import is failing on me (when trying to run /usr/bin/ops-metric-client) because 'hawkular.metrics' doesn't exist. Where is the implementation of HawkularMetricsClient?

zgalor replied:
pip install hawkular-client

unless you have a hawkular-cleint RPM, in which case you can use it.

twiest replied:
@zgalor it's ok to use pip in dev, but we don't commit any code that requires pip.

The hawkular-client RPM will only be available in the rhel container (not centos) since we only do publicly available sources for the centos container.

As a short term fix, since we're disabling the haw_client by default right now anyway, would you mind moving this import down into the constructor _after_ the check to see if the client is disabled?

joelddiaz commented on openshift_tools/monitoring/zagg_sender.py:
```diff
@@ -20,83 +20,47 @@
     """
 
-    def __init__(self, host=None, zagg_connection=None, verbose=False, debug=False):
+    def __init__(self, config_file=None, host=None, zagg_connection=None, verbose=False, debug=False):
```
Adding 'config_file' to the front of the parameter list breaks the pre-existing ops-zagg-client. Is there a reason for adding the parameter to the front?

zgalor replied:
Good point. I moved config_file to be the last param.
I also restored:
**add_zabbix_keys** to wrap **add_metric**
**add_zabbix_dynamic_item** to wrap **add_dynamic_metric**

So that existing scripts will remain working. 
After this PR is merged, I will remove those methods in https://github.com/openshift/openshift-tools/pull/1620 which makes them obsolete.

zgalor commented:
In order to avoid dependency in external hawkular-client RPM I added (with the help of @yaacov) a temporary fix to import a local mini-hawkular package in another commit.
Once we have the hawkular-client RPM ready I will remove this commit.
@joelddiaz @twiest please review and see if something is missing in order to merge this.

yaacov commented on openshift_tools/monitoring/hawk_client.py:
```diff
@@ -0,0 +1,92 @@
+    print client.add_metric(ml)
+"""
+# The hawkular client
+# Temporary import from local mini-hawkular-client to avoid dependency in external package
```
can you add a `TODO:` or `FIXME:` so it will pop up in code editors, that we need to change this once we will have an RPM ?

twiest commented:
ok to test

twiest commented:
@zgalor can you look into fixing the failing jenkins checks please?

Sorry, I meant to run that yesterday, but forgot.

We've added you and @yaacov  to the Jenkins whitelist, so it should start testing your PRs automatically.

zgalor commented:
@twiest I fixed the pylint issues. Can you please re-run tests?

twiest commented:
@zgalor as you can see from the comments, the bot is now auto-running on changes you make to this PR (and all future PRs). The bot says the tests are still failing.

Looking at the output from jenkins, and comparing it to the code in the PR, I see that Jenkins is right, there are still some pylint issues in the PR.

For instance, pylint is saying this method is missing a doc string, and it is actually missing a doc string:
https://github.com/openshift/openshift-tools/pull/1511/files#diff-de1d93fda027d0eacfec12c597c543a6R138

zgalor commented:
@twiest can we merge?

twiest commented:
@zgalor It's very close. @joelddiaz said he's going to do some final testing today.

joelddiaz reviewed (changes requested):
Need to package up the mini_hawkular files into an RPM.

Scripts fail with ImportError right now.

yaacov commented:
> Scripts fail with ImportError right now.

What scripts fail ? Can we run this tests in our environment to test before pushing to this PR ?
I'm sure we can add the calls to Hawkular using `urllib` without triggering  `ImportError`

> Need to package up the mini_hawkular files into an RPM.

@twiest @joelddiaz @zgalor :stuck_out_tongue_winking_eye:  Ahhh.... the idea of implementing Hawkular calls using `urllib` instead of `hawkular-client` is a temporary fix until you can pack an RPM for the compete hawkular client.
Packing the mini client (using urllib) does not solve our problem ... :sunglasses:

twiest commented:
@yaacov what @joelddiaz was asking is that it be added to one of our openshift tools RPMs.

The only way a library like this makes it into the container is through an RPM (either ours or external). In this case, we want it in one of our RPMs.

Specifically, I believe it should be added to this spec file:
https://github.com/openshift/openshift-tools/blob/stg/openshift_tools/python-openshift-tools.spec

To test, you can do tito test builds and ensure that your files are in the rpm.

Tito test builds are described here (specifically steps 1 and 4):
https://github.com/openshift/openshift-tools/blob/stg/docs/build.adoc#build-python-openshift-tools-rpms

Basically change directory into openshift-tools/openshift_tools (where the spec file is) and then run this:
```tito build --test --rpm --output "$HOME/.tito-build"```

And then look at the contents of the package you added it to with ```rpm -qpl <path_to_package>```

yaacov commented:
> Specifically, I believe it should be added to this spec file:
https://github.com/openshift/openshift-tools/blob/stg/openshift_tools/python-openshift-tools.spec

@twiest 👍 thanks :-)
//cc @zgalor

zgalor commented:
@twiest @joelddiaz  can I pack the mini_hawkular files inside monitoring-zagg subpackage or do I need a different subpackage for that?

joelddiaz commented:
@zgalor i'd say just throw it in there with the monitoring-zagg subpackage for now, and we'll deal with renaming things later

yaacov commented:
p.s.
@zgalor and remember to copy it to `%{buildroot}` ... :-)

zgalor commented:
@joelddiaz I added the files to the spec but I don't see them in the package after I run the tito build.
Can you see what I missed?
https://github.com/openshift/openshift-tools/pull/1511/commits/a69f53c629d3a2b210d0ff097d074dcd20c39e63#diff-d8f3d0203f20f09d4630606d180fe099

zgalor commented:
Ah! I see them now!! :-)
@joelddiaz can you please check?

joelddiaz commented:
@zgalor LGTM

zgalor commented:
Can you merge it?
@joeldiaz @twiest

joelDiaz commented:
@zgalor hey man what you do???? jejeje you are confused...

zgalor commented:
Sorry wrong @joeldiaz (no auto-complete on my phone)
The right @joelddiaz. 
:-)

joelddiaz commented:
@zgalor we try not to merge large(ish) changes on Fridays, but we can do this on Monday

joelDiaz commented:
jajaja @zgalor  dont worry man,.

zgalor commented:
@joelddiaz @twiest please wait with the merge.
I'm considering replacing the local import solution with a configuration dependent import of hawkular-client.

zgalor committed:
Elegant ImportError handling of hawkular-client

Only attempt to import hawkular-client in case hawk is configured active in metric_sendar.yaml
If import failed print error and treat as inactive but don't fail program.

zgalor commented:
I decided that handling ImportError when hawkular-client is missing is a much more elegant solution, and will require no code changes in the future, in case hawkular-client changes.

I only attempt to import hawkular-client in case hawk is configured active in metric_sender.yaml
If import failed print error and treat as inactive but don't fail program.

I tested all scenarios locally: package installed/not - hawk enabled/disabled.

cc @simon3z

@joelddiaz @twiest Would you please make a final review and merge?
